### DISCUSSION TITLE
Several questions about the `lazy.nvim` events

### DISCUSSION THREAD

**Discussion by vollowx** [ayamir/nvimdots#960] [Ideas]
1. As `lazy.nvim` provides a event `VeryLazy`, I want to know why we use `CursorHold` and `CursorHoldI` instead of that.
2. A idea, maybe we could add user custom events `File` and `GitFile` like AstroVim, in order to setup plugins more precisely.

**Comment by Jint-lzxy** [comment]
> 1. As `lazy.nvim` provides a event `VeryLazy`, I want to know why we use `CursorHold` and `CursorHoldI` instead of that.

This is very similar to #734 (which u opened before). Since:
> **VeryLazy:** triggered after `LazyDone` and processing `VimEnter` auto commands

`VeryLazy` is still triggered within the main event loop, and TBH, even worse - b/c its triggered somewhere around `UIEnter`. This will completely block the event loop, thereby delaying the time _UI_ gets rendered (but this won't affect the time calculated by `Lazy profile` tho) u can try to limit the amount of resources `nvim` can use, so that u could intuitively feel the difference. Using `CursorHold` and `CursorHoldI` is kindof like loading "extra" plugins when the editor is in a good state, just like many other editors do.

> 2. A idea, maybe we could add user custom events `File` and `GitFile` like AstroVim, in order to setup plugins more precisely.

Haven't used AstroVim before, so I can't comment on this. Maybe you can describe some possible implementation details in https://github.com/ayamir/nvimdots/pull/931?

**Comment by vollowx** [comment]
> This is very similar to https://github.com/ayamir/nvimdots/pull/734 (which u opened before). Since:

Thanks, I just forgot your explanation in that pr and can't find it.

> Haven't used AstroVim before, so I can't comment on this. Maybe you can describe some possible implementation details in https://github.com/ayamir/nvimdots/pull/931?

I will, there's a quick view.

Un-expanded code

```lua
-- File detection (File and GitFile) for lazy.nvim
{
  { 'BufReadPost', 'BufNewFile', 'BufWritePost' },
  {
    group = 'FileEvents',
    callback = function(args)
      if
        not (vim.fn.expand '%' == '' or vim.api.nvim_get_option_value('buftype', { buf = args.buf }) == 'nofile')
      then
        release_event 'File'
        if
          require('utils').git.file_worktree()
          or require('utils').command.cmd({ 'git', '-C', vim.fn.expand '%:p:h', 'rev-parse' }, false)
        then
          release_event 'GitFile'
          vim.api.nvim_del_augroup_by_name 'FileEvents'
        end
      end
    end,
  },
},
```

autocmd handler

```lua
if autocmd[2].group and vim.fn.exists('#' .. autocmd[2].group) == 0 then
  vim.api.nvim_create_augroup(autocmd[2].group, { clear = true })
end
vim.api.nvim_create_autocmd(unpack(autocmd))
```

utils.command.cmd

```lua
---Run a shell command and capture the output and if the command succeeded or failed
---@param cmd string|string[] The terminal command to execute
---@param show_error? boolean Whether or not to show an unsuccessful command as an error to the user
---@return string|nil # The result of a successfully executed command or nil
function M.cmd(cmd, show_error)
  if type(cmd) == 'string' then cmd = { cmd } end
  if vim.fn.has 'win32' == 1 then cmd = vim.list_extend({ 'cmd.exe', '/C' }, cmd) end
  local result = vim.fn.system(cmd)
  local success = vim.api.nvim_get_vvar 'shell_error' == 0
  if not success and (show_error == nil or show_error) then
    vim.api.nvim_err_writeln(('Error running command %s\nError message:\n%s'):format(table.concat(cmd, ' '), result))
  end
  return success and result:gsub('[\27\155][][()#;?%d]*[A-PRZcf-ntqry=><~]', '') or nil
end
```

utils.git.file_worktree

```lua
function M.file_worktree(file, worktrees)
  -- worktrees = worktrees or vim.g.git_worktrees
  worktrees = worktrees
  if not worktrees then return end
  file = file or vim.fn.expand '%'
  for _, worktree in ipairs(worktrees) do
    if
      vim.system({
        'git',
        '--work-tree',
        worktree.toplevel,
        '--git-dir',
        worktree.gitdir,
        'ls-files',
        '--error-unmatch',
        file,
      }, false)
    then
      return worktree
    end
  end
end
```

release_event

```lua
--- Trigger an user event
---@param event string The event name to be appended to Astro
function M.release(event)
  vim.schedule(
    function()
      vim.api.nvim_exec_autocmds('User', {
        pattern = event,
        modeline = false,
      })
    end
  )
end
```

These small feature units can be put in utils, so I did not modify them.

After all, use 'User File' event and `User GitFile` in `lazy.nvim`

**Reply by charliie-dev** [reply]
I never used `AstroVim` before, what's the scenario of useing `File` and `GitFile` events?
could u give us some examples?

**Reply by vollowx** [reply]
We're loading plugins like `gitsigns` in any files, including files that isn't inside a Git environment. Though it's actually not a big problem for performances.

**Reply by Jint-lzxy** [reply]
> We're loading plugins like `gitsigns` in any files, including files that isn't inside a Git environment. Though it's actually not a big problem for performances.

IMO having `gitsigns` is more than enough in this case. Because `gitsigns` already has the complete set of ability to handle Git workspaces - it runs at lowest cost if the workspace isn't managed by git, and also provides convenient (and asynchronous) hooks to detect whether the current buffer is within any git workspace. So adding `GitFile` doesn't sound quite useful to me (in similar cases).
